src/flint/types/fmpq_mpoly.pyx

```diff
@@ -0,0 +1,1051 @@
+            raise ValueError("Unreasonably large polynomial")  # pragma: no cover
+        return vres
+
+    def monoms(self):
+        """
+        Return the exponent vectors of each term as a tuple of fmpz.
+
+            >>> from flint import Ordering
+            >>> ctx = fmpq_mpoly_ctx.get_context(2, Ordering.lex, 'x')
+            >>> f = ctx.from_dict({(0, 0): 1, (1, 0): 2, (0, 1): 3, (1, 1): 4})
+            >>> f.keys()
+            [(1, 1), (1, 0), (0, 1), (0, 0)]
+
+        """
+        cdef:
+            slong i, nvars = self.ctx.nvars()
+            fmpz_vec vec = fmpz_vec(nvars, double_indirect=True)
+
+        res = []
+        for i in range(len(self)):
+            fmpq_mpoly_get_term_exp_fmpz(vec.double_indirect, self.val, i, self.ctx.val)
+            res.append(vec.to_tuple())
+
+        return res
+
+    def coeffs(self):
+        """
+        Return the coefficients of each term as a fmpq.
+
+            >>> from flint import Ordering
+            >>> ctx = fmpq_mpoly_ctx.get_context(2, Ordering.lex, 'x')
+            >>> f = ctx.from_dict({(0, 0): 1, (1, 0): 2, (0, 1): 3, (1, 1): 4})
+            >>> f.values()
+            [4, 2, 3, 1]
+
+        """
+        cdef:
+            fmpq coeff
+            slong i
+
+        res = []
+        for i in range(len(self)):
+            coeff = fmpq.__new__(fmpq)
+            fmpq_mpoly_get_term_coeff_fmpq(coeff.val, self.val, i, self.ctx.val)
+            res.append(coeff)
+
+        return res
+
+    def terms(self):
+        """
+        Return the terms of this polynomial as a list of fmpq_mpolys.
+
+            >>> from flint import Ordering
+            >>> ctx = fmpq_mpoly_ctx.get_context(2, Ordering.lex, 'x')
+            >>> f = ctx.from_dict({(0, 0): 1, (1, 0): 2, (0, 1): 3, (1, 1): 4})
+            >>> f.terms()
+            [4*x0*x1, 2*x0, 3*x1, 1]
+
+        """
+        cdef:
+            fmpq_mpoly term
+            slong i
+
+        res = []
+        for i in range(len(self)):
+            term = create_fmpq_mpoly(self.ctx)
+            fmpq_mpoly_get_term(term.val, self.val, i, self.ctx.val)
+            res.append(term)
+
+        return res
+
+    def subs(self, dict_args) -> fmpq_mpoly:
+        """
+        Partial evaluate this polynomial.
+        """
+        cdef:
+            fmpq_mpoly res
+            slong i, nargs
+
+        partial_args = tuple((i, dict_args[x]) for i, x in enumerate(self.ctx.names()) if x in dict_args)
+        nargs = len(partial_args)
+
+        # If we've been provided with an invalid keyword arg then the length of our filter
+        # args will be less than what we've been provided with.
+        # If the length is equal to the number of variables then all arguments have been provided
+        # otherwise we need to do partial application
+        if nargs < len(dict_args):
+            raise ValueError("Unknown keyword argument provided")
```

Comment:
I don't think this comment makes sense here any more or the error message:
```
In [69]: p.subs({0:1,1:2})
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[69], line 1
----> 1 p.subs({0:1,1:2})

File fmpq_mpoly.pyx:671, in flint.types.fmpq_mpoly.fmpq_mpoly.subs()

ValueError: Unknown keyword argument provided
```

Reply:
It certainly does not, it also currently only accepts variable names, not ints. Should that be changed?

Reply:
>  only accepts variable names, not ints.

I think it needs to be consistent across different methods. Do we expect code interacting with this to need to use variable names in some places and integers etc in others?

More generally in SymPy two different symbols can have the same name so at some point we are going to run into problems with interfaces based on names.

Reply:
Looking through the tests added it seems that `subs` is the only method that uses the string names for the variables.

Reply:
I've just allowed subs to accept generator indices as well.